File: src/R/analyse_data.R
#!/usr/bin/env Rscript

# EvoNEST Data Analysis - Table Building Script
# Loads data from EvoNEST and processes it into structured tables for analysis
# Manages configuration in config/analyse_data_config.json

# Try to load required packages, install if missing
required_packages <- c("here", "tidyverse", "jsonlite", "R6")

for (pkg in required_packages) {
  if (!requireNamespace(pkg, quietly = FALSE)) {
    cat("Installing", pkg, "...\n")
    install.packages(pkg)
  }
}

# Load libraries
suppressPackageStartupMessages({
  library(tidyverse)
  library(jsonlite)
  library(here)
  library(R6)
})

# Helper operators
`%||%` <- function(x, y) if (is.null(x)) y else x


# Configuration Manager ======================================================

ConfigManager <- R6::R6Class(
  "ConfigManager",
  public = list(
    config_file = NULL,
    config = NULL,

    initialize = function() {
      self$config_file <- here::here("config", "analyse_data_config.json")
      self$config <- self$load_config()
    },

    load_config = function() {
      default_config <- list(
        paths = list(
          downloaded_data_dir = "downloaded_data",
          processed_data_dir = "processed_data"
        ),
        output = list(
          save_tables = FALSE,
          output_format = "csv"
        )
      )

      if (file.exists(self$config_file)) {
        tryCatch({
          saved_config <- jsonlite::read_json(self$config_file)
          # Merge with defaults
          config <- private$merge_configs(default_config, saved_config)
          return(config)
        }, error = function(e) {
          cat("⚠️  Warning: Could not read config file, using defaults\n")
          return(default_config)
        })
      } else {
        return(default_config)
      }
    },

    save_config = function() {
      tryCatch({
        config_dir <- dirname(self$config_file)
        dir.create(config_dir, showWarnings = FALSE, recursive = TRUE)
        jsonlite::write_json(self$config, self$config_file, pretty = TRUE, auto_unbox = TRUE)
        cat(sprintf("✅ Configuration saved to %s\n", self$config_file))
        return(TRUE)
      }, error = function(e) {
        cat(sprintf("⚠️  Warning: Could not save config: %s\n", e$message))
        return(FALSE)
      })
    }
  ),

  private = list(
    merge_configs = function(default, saved) {
      for (key in names(saved)) {
        if (key %in% names(default) &&
            is.list(default[[key]]) && is.list(saved[[key]])) {
          default[[key]] <- private$merge_configs(default[[key]], saved[[key]])
        } else {
          default[[key]] <- saved[[key]]
        }
      }
      return(default)
    }
  )
)


# Data Loading Functions =====================================================

load_data <- function(config) {
  cat("📂 Loading data files...\n")

  # Define data paths from config
  data_dir <- here::here(config$paths$downloaded_data_dir)
  processed_dir <- here::here(config$paths$processed_data_dir)

  # Load samples data
  # Use simplifyVector = FALSE to preserve the JSON structure correctly
  samples_json <- fromJSON(file.path(data_dir, "samples_data.json"), simplifyVector = FALSE)
  cat(sprintf("  ✓ Loaded %d samples\n", length(samples_json$samples)))

  # Load traits data
  traits_json <- fromJSON(file.path(data_dir, "traits_data.json"), simplifyVector = FALSE)
  cat(sprintf("  ✓ Loaded %d traits\n", length(traits_json$traits)))

  # Load processed experiments data
  experiments_json <- fromJSON(file.path(processed_dir, "hierarchical_experiment_data_no_curves.json"), simplifyVector = FALSE)
  cat(sprintf("  ✓ Loaded %d experiments\n",
              length(experiments_json$experiments)))

  cat("\n")

  return(list(
    samples = samples_json,
    traits = traits_json,
    experiments = experiments_json
  ))
}


# Table Building Functions ===================================================

build_samples_table <- function(samples_json) {
  cat("🔨 Building samples table...\n")

  # Convert samples list to data frame
  samples_df <- samples_json$samples %>%
    map_df(~{
      # Remove complex nested structures that can't be flattened
      if ("logbook" %in% names(.x)) {
        .x[["logbook"]] <- NULL  # Remove for now, too complex
      }
      if ("secondaryItems" %in% names(.x)) {
        .x[["secondaryItems"]] <- NULL
      }
      if ("filesId" %in% names(.x)) {
        .x[["filesId"]] <- NULL
      }
      
      # Convert all single-element lists to atomic values
      # This handles fields like "type": ["animal"] -> "type": "animal"
      .x <- lapply(.x, function(field) {
        if (is.list(field) && length(field) == 1 && !is.list(field[[1]])) {
          return(field[[1]])
        }
        return(field)
      })
      
      # Return as tibble
      as_tibble(.x)
    })

  cat(sprintf("  ✓ Samples DataFrame: %d rows × %d columns\n",
              nrow(samples_df), ncol(samples_df)))

  # Display sample types distribution
  if ("type" %in% colnames(samples_df)) {
    sample_types <- table(samples_df$type)
    cat("  Sample types:\n")
    for (type_name in names(sample_types)) {
      cat(sprintf("    - %s: %d\n", type_name, sample_types[type_name]))
    }
  }

  cat("\n")
  return(samples_df)
}


build_traits_table <- function(traits_json) {
  cat("🔨 Building traits table...\n")

  # Convert traits list to data frame
  traits_df <- traits_json$traits %>%
    map_df(~{
      # Flatten nested sample data
      if ("sample" %in% names(.x) && !is.null(.x[["sample"]])) {
        sample_data <- .x[["sample"]]
        .x[["sample"]] <- NULL
        # Add selected sample fields with prefix - use NA_character_ for consistent typing
        .x[["sample.name"]] <- sample_data[["name"]] %||% NA_character_
        .x[["sample.type"]] <- sample_data[["type"]] %||% NA_character_
        .x[["sample.family"]] <- sample_data[["family"]] %||% NA_character_
        .x[["sample.genus"]] <- sample_data[["genus"]] %||% NA_character_
        .x[["sample.species"]] <- sample_data[["species"]] %||% NA_character_
        .x[["sample.nomenclature"]] <- sample_data[["nomenclature"]] %||% NA_character_
        .x[["sample.subsampletype"]] <- sample_data[["subsampletype"]] %||% NA_character_
      }
      
      # Remove all complex nested structures that might cause issues
      fields_to_remove <- c("logbook", "filesId", "secondaryItems", "diameterConversion")
      for (field in fields_to_remove) {
        if (field %in% names(.x)) {
          .x[[field]] <- NULL
        }
      }
      
      # Handle listvals - convert to string representation
      if ("listvals" %in% names(.x) && !is.null(.x[["listvals"]])) {
        listvals_data <- .x[["listvals"]]
        # Check if it's an empty object (named list with no elements or all NULL)
        if (is.list(listvals_data) && (length(listvals_data) == 0 || !is.null(names(listvals_data)))) {
          # Empty object or named list - set to empty string
          .x[["listvals"]] <- ""
        } else if (is.list(listvals_data) && length(listvals_data) > 0) {
          # Convert array to comma-separated string
          vals <- unlist(listvals_data)
          .x[["listvals"]] <- paste(as.character(vals), collapse = ", ")
        } else {
          .x[["listvals"]] <- ""
        }
      }
      
      # Convert all single-element lists to atomic values
      .x <- lapply(.x, function(field) {
        if (is.list(field) && length(field) == 1 && !is.list(field[[1]])) {
          return(field[[1]])
        }
        return(field)
      })
      
      # Remove any remaining nested list/object fields that aren't simple values
      .x <- .x[sapply(.x, function(field) {
        # Keep if it's not a list, or if it's a character/numeric/logical vector
        !is.list(field) || is.null(field)
      })]
      
      .x
    })

  cat(sprintf("  ✓ Traits DataFrame: %d rows × %d columns\n",
              nrow(traits_df), ncol(traits_df)))

  # Display trait types distribution
  if ("type" %in% colnames(traits_df)) {
    trait_types <- table(traits_df$type)
    cat(sprintf("  Trait types: %d unique types\n", length(trait_types)))
    # Show top 5 trait types
    top_traits <- sort(trait_types, decreasing = TRUE)[1:min(5, length(trait_types))]
    cat("  Top trait types:\n")
    for (i in seq_along(top_traits)) {
      cat(sprintf("    - %s: %d\n", names(top_traits)[i], top_traits[i]))
    }
  }

  cat("\n")
  return(traits_df)
}


build_experiments_table <- function(experiments_json) {
  cat("🔨 Building experiments table...\n")

  # Helper function to extract value from potential single-element list
  extract_value <- function(x) {
    if (is.list(x) && length(x) == 1 && !is.list(x[[1]])) {
      return(x[[1]])
    }
    return(x)
  }

  # Convert experiments nested list to data frame
  experiments_list <- names(experiments_json$experiments) %>%
    map_df(~{
      exp_data <- experiments_json$experiments[[.x]]

      # Create base record with flattened data
      exp_record <- tibble(
        experiment_id = .x,
        sample_name = extract_value(exp_data$sample_name) %||% NA,
        type = extract_value(exp_data$type) %||% NA,
        date = extract_value(exp_data$date) %||% NA,
        r_squared = extract_value(exp_data$r_squared) %||% NA,
        data_points = extract_value(exp_data$data_points) %||% NA,
        fracture_detected = extract_value(exp_data$fracture_detected) %||% NA,
        max_stress = extract_value(exp_data$max_stress) %||% NA,
        responsible = extract_value(exp_data$responsible) %||% NA,
        notes = extract_value(exp_data$notes) %||% NA,
        equipment = extract_value(exp_data$equipment) %||% NA,
        family = extract_value(exp_data$family) %||% NA,
        genus = extract_value(exp_data$genus) %||% NA,
        species = extract_value(exp_data$species) %||% NA,
        subsampletype = extract_value(exp_data$subsampletype) %||% NA
      )

      # Add strain and stress ranges
      strain_range <- extract_value(exp_data$strain_range)
      if (!is.null(strain_range) && length(strain_range) >= 2) {
        exp_record$strain_min <- strain_range[[1]]
        exp_record$strain_max <- strain_range[[2]]
      }
      stress_range <- extract_value(exp_data$stress_range)
      if (!is.null(stress_range) && length(stress_range) >= 2) {
        exp_record$stress_min <- stress_range[[1]]
        exp_record$stress_max <- stress_range[[2]]
      }

      # Store polynomial coefficients as list column
      poly_coeffs <- extract_value(exp_data$polynomial_coefficients)
      if (is.list(poly_coeffs) && !is.null(poly_coeffs)) {
        exp_record$polynomial_coefficients <- list(unlist(poly_coeffs))
      } else {
        exp_record$polynomial_coefficients <- list(poly_coeffs)
      }

      exp_record
    })

  experiments_df <- experiments_list

  cat(sprintf("  ✓ Experiments DataFrame: %d rows × %d columns\n",
              nrow(experiments_df), ncol(experiments_df)))

  cat("\n")
  return(experiments_df)
}


# Summary Functions ==========================================================

print_summary <- function(samples_df, traits_df, experiments_df) {
  cat(strrep("=", 80), "\n")
  cat("DATA SUMMARY\n")
  cat(strrep("=", 80), "\n\n")

  cat("📊 SAMPLES\n")
  cat(sprintf("  Total samples: %d\n", nrow(samples_df)))
  if ("type" %in% colnames(samples_df)) {
    sample_counts <- table(samples_df$type)
    cat("  Sample types:\n")
    for (type_name in names(sample_counts)) {
      cat(sprintf("    - %s: %d\n", type_name, sample_counts[type_name]))
    }
  }
  if ("family" %in% colnames(samples_df)) {
    cat(sprintf("  Families represented: %d\n",
                length(unique(samples_df$family[!is.na(samples_df$family)]))))
  }

  cat("\n🔬 TRAITS\n")
  cat(sprintf("  Total traits: %d\n", nrow(traits_df)))
  if ("type" %in% colnames(traits_df)) {
    cat(sprintf("  Trait types: %d unique types\n",
                length(unique(traits_df$type[!is.na(traits_df$type)]))))
  }

  cat("\n⚗️ EXPERIMENTS\n")
  cat(sprintf("  Total experiments: %d\n", nrow(experiments_df)))
  if ("r_squared" %in% colnames(experiments_df)) {
    cat(sprintf("  Average R²: %.4f\n",
                mean(experiments_df$r_squared, na.rm = TRUE)))
  }
  if ("fracture_detected" %in% colnames(experiments_df)) {
    cat(sprintf("  Fracture detected: %d / %d\n",
                sum(experiments_df$fracture_detected, na.rm = TRUE),
                nrow(experiments_df)))
  }
  if ("family" %in% colnames(experiments_df)) {
    cat(sprintf("  Families tested: %d\n",
                length(unique(experiments_df$family[!is.na(experiments_df$family)]))))
  }

  cat("\n")
  cat(strrep("=", 80), "\n")
}


# Main Function ==============================================================

main <- function() {
  cat("\n", strrep("═", 80), "\n", sep = "")
  cat("EvoNEST Data Analysis - Building Data Tables\n")
  cat(strrep("═", 80), "\n\n", sep = "")

  # Load configuration
  config_manager <- ConfigManager$new()
  config <- config_manager$config

  # Load data
  data <- load_data(config)

  # Build tables
  samples_df <- build_samples_table(data$samples)
  traits_df <- build_traits_table(data$traits)
  experiments_df <- build_experiments_table(data$experiments)

  # Print summary
  print_summary(samples_df, traits_df, experiments_df)

  cat("\n✅ Data tables built successfully!\n")
  cat("   Available objects: samples_df, traits_df, experiments_df\n")
  cat("\n💡 Next steps: Explore data and create visualizations with ggplot2\n")
  cat(strrep("═", 80), "\n\n", sep = "")

  # Assign to global environment so they're available after sourcing
  assign("samples_df", samples_df, envir = .GlobalEnv)
  assign("traits_df", traits_df, envir = .GlobalEnv)
  assign("experiments_df", experiments_df, envir = .GlobalEnv)

  # Return results invisibly
  invisible(list(
    samples_df = samples_df,
    traits_df = traits_df,
    experiments_df = experiments_df
  ))
}


# Run main function
main()

File: src/python/analyse_data.py
#!/usr/bin/env python3

"""
EvoNEST Data Analysis - Table Building Script
Loads data from EvoNEST and processes it into structured tables for analysis
Manages configuration in config/analyse_data_config.json
"""

import json
from pathlib import Path
from typing import Dict, Tuple

import pandas as pd


class ConfigManager:
    """Manages persistent configuration for data analysis"""

    CONFIG_DIR = Path(__file__).parent.parent / "config"
    CONFIG_FILE = CONFIG_DIR / "analyse_data_config.json"

    DEFAULT_CONFIG = {
        "paths": {
            "downloaded_data_dir": "downloaded_data",
            "processed_data_dir": "processed_data"
        },
        "output": {
            "save_tables": False,
            "output_format": "csv"
        }
    }

    def __init__(self):
        """Initialize config manager"""
        self.config_dir = self.CONFIG_DIR
        self.config_file = self.CONFIG_FILE
        self.config = self.load_config()

    def load_config(self) -> Dict:
        """Load configuration from file or create default"""
        if self.config_file.exists():
            try:
                with open(self.config_file, 'r', encoding='utf-8') as f:
                    saved_config = json.load(f)
                # Merge with defaults
                config = self._merge_configs(self.DEFAULT_CONFIG.copy(), saved_config)
                return config
            except (json.JSONDecodeError, IOError):
                print("⚠️  Warning: Could not read config file, using defaults")
                return self.DEFAULT_CONFIG.copy()
        else:
            return self.DEFAULT_CONFIG.copy()

    def save_config(self) -> bool:
        """Save configuration to file"""
        try:
            self.config_dir.mkdir(parents=True, exist_ok=True)
            with open(self.config_file, 'w', encoding='utf-8') as f:
                json.dump(self.config, f, indent=2)
            print(f"✅ Configuration saved to {self.config_file}")
            return True
        except IOError as e:
            print(f"⚠️  Warning: Could not save config: {e}")
            return False

    def _merge_configs(self, default: Dict, saved: Dict) -> Dict:
        """Recursively merge saved config with defaults"""
        result = default.copy()
        for key, value in saved.items():
            if key in result and isinstance(result[key], dict) and isinstance(value, dict):
                result[key] = self._merge_configs(result[key], value)
            else:
                result[key] = value
        return result


def load_data(config: Dict) -> Tuple[Dict, Dict, Dict]:
    """Load all data files from downloaded_data and processed_data folders."""
    print("📂 Loading data files...")

    # Define data paths from config
    base_dir = Path(__file__).parent.parent
    data_dir = base_dir / config['paths']['downloaded_data_dir']
    processed_dir = base_dir / config['paths']['processed_data_dir']

    # Load samples data
    with open(data_dir / 'samples_data.json', 'r', encoding='utf-8') as f:
        samples_json = json.load(f)
    print(f"  ✓ Loaded {len(samples_json['samples'])} samples")

    # Load traits data
    with open(data_dir / 'traits_data.json', 'r', encoding='utf-8') as f:
        traits_json = json.load(f)
    print(f"  ✓ Loaded {len(traits_json['traits'])} traits")

    # Load processed experiments data
    with open(processed_dir / 'hierarchical_experiment_data_no_curves.json', 'r', encoding='utf-8') as f:
        experiments_json = json.load(f)
    print(f"  ✓ Loaded {len(experiments_json['experiments'])} experiments")

    print()
    return samples_json, traits_json, experiments_json


def build_samples_table(samples_json):
    """Build samples DataFrame from JSON data."""
    print("🔨 Building samples table...")
    
    # Convert to DataFrame using json_normalize
    samples_df = pd.json_normalize(samples_json['samples'])
    
    print(f"  ✓ Samples DataFrame: {samples_df.shape[0]} rows × {samples_df.shape[1]} columns")
    
    # Display sample types distribution
    if 'type' in samples_df.columns:
        sample_types = samples_df['type'].value_counts()
        print("  Sample types:")
        for type_name, count in sample_types.items():
            print(f"    - {type_name}: {count}")
    
    print()
    return samples_df


def build_traits_table(traits_json):
    """Build traits DataFrame from JSON data."""
    print("🔨 Building traits table...")
    
    # Convert to DataFrame using json_normalize
    traits_df = pd.json_normalize(traits_json['traits'])
    
    print(f"  ✓ Traits DataFrame: {traits_df.shape[0]} rows × {traits_df.shape[1]} columns")
    
    # Display trait types distribution
    if 'type' in traits_df.columns:
        trait_types = traits_df['type'].value_counts()
        print(f"  Trait types: {len(trait_types)} unique types")
        print("  Top trait types:")
        for type_name, count in trait_types.head(5).items():
            print(f"    - {type_name}: {count}")
    
    print()
    return traits_df


def build_experiments_table(experiments_json):
    """Build experiments DataFrame from processed data."""
    print("🔨 Building experiments table...")
    
    # Convert experiments dict to list of records
    experiments_list = []
    for exp_id, exp_data in experiments_json['experiments'].items():
        # Flatten the nested structure
        exp_record = {
            'experiment_id': exp_id,
            'sample_name': exp_data.get('sample_name'),
            'type': exp_data.get('type'),
            'date': exp_data.get('date'),
            'r_squared': exp_data.get('r_squared'),
            'data_points': exp_data.get('data_points'),
            'fracture_detected': exp_data.get('fracture_detected'),
            'max_stress': exp_data.get('max_stress'),
            'responsible': exp_data.get('responsible'),
            'notes': exp_data.get('notes'),
            'equipment': exp_data.get('equipment'),
            'family': exp_data.get('family'),
            'genus': exp_data.get('genus'),
            'species': exp_data.get('species'),
            'subsampletype': exp_data.get('subsampletype'),
        }
        
        # Add strain and stress ranges
        if 'strain_range' in exp_data:
            exp_record['strain_min'] = exp_data['strain_range'][0]
            exp_record['strain_max'] = exp_data['strain_range'][1]
        
        if 'stress_range' in exp_data:
            exp_record['stress_min'] = exp_data['stress_range'][0]
            exp_record['stress_max'] = exp_data['stress_range'][1]
        
        # Store polynomial coefficients as list
        exp_record['polynomial_coefficients'] = exp_data.get('polynomial_coefficients')
        
        experiments_list.append(exp_record)
    
    experiments_df = pd.DataFrame(experiments_list)
    
    print(f"  ✓ Experiments DataFrame: {experiments_df.shape[0]} rows × {experiments_df.shape[1]} columns")
    print()
    
    return experiments_df


def print_summary(samples_df, traits_df, experiments_df):
    """Print summary statistics for all tables."""
    print("=" * 80)
    print("DATA SUMMARY")
    print("=" * 80)
    print()
    
    print("📊 SAMPLES")
    print(f"  Total samples: {len(samples_df)}")
    if 'type' in samples_df.columns:
        sample_counts = samples_df['type'].value_counts()
        print("  Sample types:")
        for type_name, count in sample_counts.items():
            print(f"    - {type_name}: {count}")
    if 'family' in samples_df.columns:
        print(f"  Families represented: {samples_df['family'].nunique()}")
    
    print("\n🔬 TRAITS")
    print(f"  Total traits: {len(traits_df)}")
    if 'type' in traits_df.columns:
        print(f"  Trait types: {traits_df['type'].nunique()} unique types")
    
    print("\n⚗️ EXPERIMENTS")
    print(f"  Total experiments: {len(experiments_df)}")
    if 'r_squared' in experiments_df.columns:
        print(f"  Average R²: {experiments_df['r_squared'].mean():.4f}")
    if 'fracture_detected' in experiments_df.columns:
        print(f"  Fracture detected: {experiments_df['fracture_detected'].sum()} / {len(experiments_df)}")
    if 'family' in experiments_df.columns:
        print(f"  Families tested: {experiments_df['family'].nunique()}")
    
    print()
    print("=" * 80)


def main() -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
    """Main function to build all data tables."""
    print("\n" + "═" * 80)
    print("EvoNEST Data Analysis - Building Data Tables")
    print("═" * 80)
    print()

    # Load configuration
    config_manager = ConfigManager()
    config = config_manager.config

    # Load data
    samples_json, traits_json, experiments_json = load_data(config)

    # Build tables
    samples_df = build_samples_table(samples_json)
    traits_df = build_traits_table(traits_json)
    experiments_df = build_experiments_table(experiments_json)

    # Print summary
    print_summary(samples_df, traits_df, experiments_df)

    print("\n✅ Data tables built successfully!")
    print("   Available DataFrames: samples_df, traits_df, experiments_df")
    print("\n💡 Next steps: Explore data and create visualizations with seaborn/matplotlib")
    print("═" * 80 + "\n")

    return samples_df, traits_df, experiments_df


if __name__ == "__main__":
    samples_df, traits_df, experiments_df = main()
